contracts/FlowIDTableStaking.cdc

```diff
@@ -870,6 +870,79 @@ pub contract FlowIDTableStaking {
+                ?? panic("Could not load the slot limits object from storage!")
+
+            // count the nodes of each type
+            var nodeCounts: {UInt8: UInt8} = {1: 0, 2: 0, 3: 0, 4: 0, 5: 0}
+            for nodeID in allNodeIDs {
+                let nodeRecord = FlowIDTableStaking.borrowNodeRecord(nodeID)
+                if nodeRecord.initialWeight == 100 {
+                    nodeCounts[nodeRecord.role] = nodeCounts[nodeRecord.role]! + 1
+                }
+            }
+
+            for key in nodeCounts.keys {
+                while nodeCounts[key]! > slotLimits[key]! {
+                    let selection = unsafeRandom() % UInt64(nodeCounts[key]!)
+                    var searchIndex: UInt64 = 0
+                    for nodeID in allNodeIDs {
+                        let nodeRecord = FlowIDTableStaking.borrowNodeRecord(nodeID)
+                        if nodeRecord.role == key && nodeRecord.initialWeight == 100 {
+                            if selection == searchIndex {
+                                emit NodeRemovedAndRefunded(nodeID: nodeRecord.id, amount: nodeRecord.tokensCommitted.balance + nodeRecord.tokensStaked.balance)
+
+                                // move their committed tokens back to their unstaked tokens
+                                nodeRecord.tokensUnstaked.deposit(from: <-nodeRecord.tokensCommitted.withdraw(amount: nodeRecord.tokensCommitted.balance))
+
+                                // Set their request to unstake equal to all their staked tokens
+                                // since they are forced to unstake
+                                nodeRecord.tokensRequestedToUnstake = nodeRecord.tokensStaked.balance
+
+                                // Iterate through all delegators and unstake their tokens
+                                // since their node has unstaked
+                                for delegator in nodeRecord.delegators.keys {
+                                    let delRecord = nodeRecord.borrowDelegatorRecord(delegator)
+
+                                    if delRecord.tokensCommitted.balance > 0.0 {
+                                        emit DelegatorTokensUnstaked(nodeID: nodeRecord.id, delegatorID: delegator, amount: delRecord.tokensCommitted.balance)
+
+                                        // move their committed tokens back to their unstaked tokens
+                                        delRecord.tokensUnstaked.deposit(from: <-delRecord.tokensCommitted.withdraw(amount: delRecord.tokensCommitted.balance))
+                                    }
+
+                                    // Request to unstake all tokens
+                                    delRecord.tokensRequestedToUnstake = delRecord.tokensStaked.balance
+                                }
+
+                                // Clear initial weight because the node is not staked any more
+                                nodeRecord.initialWeight = 0
```

Comment:
This (lines 906-932) looks similar to [a part of `removeUnapprovedNodes`](https://github.com/onflow/flow-core-contracts/blob/master/contracts/FlowIDTableStaking.cdc#L838-L864). If possible, it would be great to encapsulate the shared logic into a shared helper function.

Reply:
I think it may be possible to actually just merge these two functions together so that all of the remove logic happens after slot selection.

Reply:
As noted above if we merge the two functions I think we should consider changing the name. Are you envisioning that we would pre-select the nodes that would be removed by removeUnapprovedNodes and then pre-select those that would be removed by slot selection and then remove them all in one final loop?